tests/test_result_serialisation.py

```diff
@@ -0,0 +1,42 @@
+import unittest
+
+from tests.utils import TestHarness, tasks, TaskErrorException
+
+
+class MyClass:
+    def __init__(self, val):
+        self._my_field = val
+
+    @property
+    def my_field(self):
+        return self._my_field
+
+
+@tasks.bind(content_type='application/python-pickle')
+def _return_my_class_instance(inner_val):
+    return MyClass(inner_val)
+
+
+@tasks.bind(content_type='application/json')
+def _return_my_class_instance_as_json(inner_val):
+    return MyClass(inner_val)
+
+
+class ResultSerialisationTests(unittest.TestCase):
+    def setUp(self) -> None:
+        self.test_harness = TestHarness()
+
+    def test_returning_non_json_serialisable_type(self):
+        pipeline = tasks.send(_return_my_class_instance, 'my_val')
+        result = self.test_harness.run_pipeline(pipeline)
+        self.assertIsInstance(result, MyClass)
+        self.assertEqual(result.my_field, 'my_val')
+
+    def test_returning_non_json_serialisable_type_with_content_type_application_json(self):
+        pipeline = tasks.send(_return_my_class_instance_as_json, 'my_val')
+        try:
+            self.test_harness.run_pipeline(pipeline)
+        except TaskErrorException as e:
+            self.assertIn('Object of type MyClass is not JSON serializable', str(e))
+        else:
+            self.fail('Expected an exception')
```

Comment:
I noticed when writing this test that you can't call the set method to set a content_type on a task. I don't think that should be supported because the content type is dependent on the method rather than how it is called, but should we be explicit with what can be set and throw an exception otherwise?

Reply:
calling fn.set(content_type='pickle') means something different to decorating a task with @task(content_type='pickle')

The former means you are telling the task API that you are going to send data that must be pickled (rather than json serialised) and the latter means that the task expects the API to pickle its result.

Reply:
It's completely valid to send pickled data to a task that returns json formatted data and vice versa.  To the API user it should be pretty transparent.

Reply:
I ran some further tests and with a pipeline constructed server-side it works pretty much as you say. The only difference I could see is that the decorator @task(content_type='pickle') would mean both the input and output are pickled by default. But you can override the input serialisation with .set.

If you create the pipeline client side then a content_type specified in the .set method appears to be ignored. This test passes when I would expect an exception:
`
@tasks.bind(content_type='application/python-pickle')
def return_my_class_instance(inner_val):
    return MyClass(inner_val)

def test_sending_data_as_json_client_side_when_not_json_serialisable(self):
    pipeline = tasks.send(return_my_class_instance, MyClass('a')).set(content_type='application/json')
    result = self.test_harness.run_pipeline(pipeline)
`

I'll dig into this a bit further...

Reply:
Ah this is because the whole pipeline is being serialised (first attempt is json then falling back to pickle)

so 'tasks.send(return_my_class_instance, MyClass('a')).set(content_type='application/json')' - > all of this is pickled and sent across.

Reply:
but you are right that if it is a single task it should fail - line 328 of _pipeline.py - because single tasks get invoked directly not sent as a single element pipeline to __builtins.run_pipeline.  See line 318.

This is because if you are trying to invoke a single task from the client then it would be a bit odd to be getting a task_id back which would be the task_id not for your task but the pipeline which contains one task....so isn't really a pipeline!

Reply:
I see. 

It works as expected if I add a .continue_with call to make it a pipeline rather than a single task.

Reply:
yep.  I'll correct in the next release